### DISCUSSION TITLE
Timing issue with machine.deepsleep

### DISCUSSION THREAD

**Discussion by mu88** [micropython/micropython#12919] [RP2040 / Pico]
Hey everyone,

I'm facing a strange issue: after calling `machine.deepsleep(300000)` (sleep for 5 min), the Pico awakes after approx. 3 min.

I also checked `machine.wake_reason()` which returns `3`.

Is there any explanation why the Pico wakes up two minutes early?

[Here's the complete code if necessary.](https://github.com/mu88/ScreenshotCreator/blob/main/src/WaveshareDisplayCode/main.py)

Thank you for your help!

**Comment by mu88** [comment]
Interesting update: I've doubled the sleep time, i. e. now calling `machine.deepsleep(600000)` - but the Pico still wakes up after 3 min 😮

**Comment by robert-hh** [comment]
Looking at the code, the sleep time should range up to 2**32/1000, which is ~4200 seconds. Is WiFi enabled in your tests?
edit: The way it's implemented, deepsleep() is lightsleep() followed by a reset.

**Reply by mu88** [reply]
Thx for your reply! 4200 sec = 70 min, i. e. way more than my 5 min, strange. Yes, WiFi is enabled (please [see here](https://github.com/mu88/ScreenshotCreator/blob/main/src/WaveshareDisplayCode/main.py#L186-L205))

Maybe it's also worth mentioning that I just recently switched from `machine.lightsleep` to `machine.deepsleep` ([see this commit](https://github.com/mu88/ScreenshotCreator/commit/c374d89dbd3ec8ff666bc3b113ac5daa18cfc2c5)) to consume less power. Interestingly, I never had this issue with `machine.lightsleep`

**Reply by robert-hh** [reply]
There should not be any difference between deepsleep() and lightsleep(), except that deepsleep() resets the device. See the code of deepsleep() below. You could try to repeat the test a few times in one hour distance, to see if it is an overflow problem.
```
STATIC mp_obj_t machine_deepsleep(size_t n_args, const mp_obj_t *args) {
    machine_lightsleep(n_args, args);
    return machine_reset();
}
```
P.S.: It worked here with `deepsleep(300_000)`

**Reply by robert-hh** [reply]
You should call wifi.deinit() before machine.deepsleep().

**Reply by mu88** [reply]
I will try both calling `wifi.deinit` and using the other syntax (using `300_000` instead of `300000`) and report back here.

But just for my understanding: you say there should be no difference in power consumption between light and deep sleep. But according to the [docs](https://datasheets.raspberrypi.org/pico/pico-datasheet.pdf), there is the SLEEP and the DORMANT mode. Does that mean that MicroPython does only support using the Pico's SLEEP mode, but not the DORMANT mode?

**Reply by robert-hh** [reply]
As far as I understand the code, the dormant mode is used when you call lightsleep() or deepsleep() without an argument. Still calling  lightsleep() or deepsleep() runs the same code in MicroPython. The equivalent of `machine.deepsleep(n)` is `machine.lightsleep(n);machine.reset()`.

P.S.: Using 300000 or 300_000 makes no difference. It's just two different ways to write an integer.

**Reply by mu88** [reply]
I gave calling `network.WLAN(network.STA_IF).deinit()` a try yesterday evening. Unfortunately, it makes the problem even worse :( instead of sleeping for 5 min, the first wakeup happens after 2 min and the next one after 1.5 min

**Reply by robert-hh** [reply]
The deinit() does not affect the lightsleep() timing issue. It affects the current during lightsleep and reduces it from 13mA to <1mA. I still cannot reproduce that case here. Both lightsleep() and deepsleep() stop after exactly 300 seconds, if called with 300_000 as argument. With both a Pico and a Pico W.

**Reply by mu88** [reply]
It's getting even more strange. I removed the attached ePaper display from the Pico and reduced my code to this:
```python
import machine

while(True):
    print('Sleeping')
    machine.lightsleep(30_000)
```
After starting the code, the output `Sleeping` occurred every 200 ms.

So I reset the Pico's flash memory ([see here](https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html#resetting-flash-memory)), flashed again to `v.1.21.0` and started the same script. Now the output `Sleeping` occurs exactly once and never again.

Any idea what's going on here?

**Reply by robert-hh** [reply]
Lightsleep at RP2 has a problem with USB. During lightsleep, it shuts off USB, and sometimes it does not reappear again, until it receives e.g. an enter character from the intput. But your are right. machine.lightsleep() seems broken. Looking at the current consumption, I can see that the code below works. Only after some cycles the output stops.
```
import machine
import time

while(True):
    print('Sleeping')
    time.sleep_ms(50)
    machine.lightsleep(5_000)
    time.sleep_ms(50)
    print('\nAwake')
    time.sleep_ms(3000)
```

**Reply by mu88** [reply]
Unfortunately, your code doesn't work for me (neither on `v.1.21.0` nor on `v.1.20.0`): it stops immediately after `machine.lightsleep(5_000)`, I never see the output `Awake`.

So now I have the following situation: both `machine.lightsleep` and `machine.deepsleep` seem to have timing issues on `v.1.20.0` and `v.1.21.0`. Therefore, I tried using the following code instead of `machine.<deep|light>sleep` on `v.1.21.0`:
```python
network.WLAN(network.STA_IF).deinit()
time.sleep(sleep_in_seconds)
machine.reset()
```
With this, my code fails somewhere else with an `OutOfMemory` exception - which makes sense as I had a very hard time trimming my code so that I don't get this exception with previous MicroPython versions. Since `v.1.21.0` is slightly bigger (according to the [release notes](https://github.com/micropython/micropython/releases/tag/v1.21.0)), I assume this little bigger footage of MicroPython is simply too much for my code.

That's why I gave `v.1.20.0` a try with the previous `time.sleep` approach and at least it works. However, the minimum power consumption after `network.WLAN(network.STA_IF).deinit()` is now 24 mA.

**Reply by robert-hh** [reply]
If power consumption matters, you could deinit WiFi and drop the machine.freq() to 18MHz before calling time.sleep(). That drops the current consumption to ~5 mA.

**Reply by mu88** [reply]
Thx for that hint - I'm already deinit-int the WiFi. Does dropping the machine frequency have an impact on the timing? And do you know whether there are already existing issues tracking the broken sleep APIs?

**Reply by robert-hh** [reply]
A short test showed, that time.sleep() still works right at 18 MHz. About earlier issues: A short search revealed a discussion I had at issue #9006. Issue #8994 may may also be interesting.

**Reply by mu88** [reply]
#9006 looks indeed very similar to my problem - so I have to find a way to trim MicroPython (i. e. remove modules I don't need) as soon as the next version of MicroPython gets released to get rid of the `OutOfMemory` problem

**Reply by mu88** [reply]
So I can confirm that calling `machine.freq(18000000)` reduces the power consumption from 24 mA to 15 mA
